src/middleware/process.ts

```diff
@@ -1,82 +1,56 @@
-      const isLastMiddleware = middlewareIndex === (middleware.length - 1);
-      const nextWhenNotLast = isLastMiddleware ? noop : next;
+  for (const fn of middleware) {
+    if (typeof fn !== 'function') {
+      throw new TypeError('Middleware must be composed of functions!');
+    }
+  }
 
-      // In this condition, errorOrPostProcess will be a postProcess function or undefined
-      postProcessFns[middlewareIndex - 1] = errorOrPostProcess === undefined ? noopPostProcess : errorOrPostProcess;
-      thisMiddleware({ context, next: nextWhenNotLast, ...initialArguments });
+  return function (context: ProcessMiddlewareContext<AnyMiddlewareArgs>): Promise<unknown> {
+    // last called middleware #
+    let index = -1;
 
-      if (isLastMiddleware) {
-        postProcessFns[middlewareIndex] = noopPostProcess;
-        process.nextTick(next);
+    async function dispatch(order: number): Promise<unknown> {
+      if (order < index) {
+        return Promise.reject(new Error('next() called multiple times'));
       }
-      return;
-    }
-
-    // Processing is complete, and we should begin bubbling up
-    // there's no next middleware or the argument is an error
 
-    function createDone(initialIndex: number): (error?: Error) => void {
-      let postProcessIndex = initialIndex;
+      index = order;
 
-      // done is a function that handles bubbling up in a similar way to next handling propogating down
-      const done = (error?: Error): void => {
-        postProcessIndex -= 1;
+      let fn: Middleware<AnyMiddlewareArgs> | undefined  = middleware[order];
 
-        const thisPostProcess = postProcessFns[postProcessIndex];
+      if (order === middleware.length) {
+        fn = context.next;
+      }
 
-        if (thisPostProcess !== undefined) {
-          thisPostProcess(error, done);
-          return;
-        }
+      if (fn === null || fn === undefined) {
+        return;
+      }
 
-        afterPostProcess(error);
-      };
-      return done;
-    }
+      context.next = dispatch.bind(null, order + 1);
 
-    if (thisMiddleware === undefined) {
-      afterMiddleware(context, initialArguments, (error?: Error) => {
-        createDone(middleware.length)(error);
-      });
-    } else {
-      createDone(middlewareIndex - 1)(errorOrPostProcess as Error);
+      return fn((context as MiddlewareContext<AnyMiddlewareArgs>));
     }
 
+    return dispatch(0);
   };
-
-  const firstMiddleware = middleware[0];
-  firstMiddleware({ context, next, ...initialArguments });
 }
 
-function noop(): void { } // tslint:disable-line:no-empty
-const noopPostProcess: PostProcessFn = (error, done) => { done(error); };
+export async function processMiddleware(
+    middleware: Middleware<AnyMiddlewareArgs>[],
+    context: ProcessMiddlewareContext<AnyMiddlewareArgs>,
+): Promise<unknown> {
+  return composeMiddleware(middleware)({
+    ...context,
+    next: () => Promise.resolve(),
```

Comment:
An interesting thing to note here, it became quickly obvious to me why koa-compose chose to have their middleware signatures be `(ctx, next)` instead of just `(ctx)` with `next` inside `ctx`. 

The reason here is that one of the purposes of ctx is to be passed around and modified in place, it's very awkward to do that when the processing has to modify the context mid stream. (As next needs to _basically_ be the next function in the chain)

That lead to this object spread here. This will cause some (possibly) weirdness when it comes to listener middleware. Each listener middleware will have it's own context separate from the rest of them... sometimes. It would depend on the shape of context and how it was accessed and changed in processing. 

I can see this being sorta helpful (partial isolation of middleware running at the same time), but could lead to unexpected results if folks are used to writing koa style middleware.

For this reason I'd still recommend using koa-style middleware rather than this one. But, if pseudo-backwards-compatibility is still more important, this works for the most part.

Reply:
Correction, not sure why this didn't click before, _because_ all of the listener middleware is run at once, in order for them not to be constantly overriding each other's `next` function, this _must_ be spread. 

Changing the signature would let us not have to do this.